File: Classes/PBCommitList.m
//
//  PBCommitList.m
//  GitX
//
//  Created by Pieter de Bie on 9/11/08.
//  Copyright 2008 __MyCompanyName__. All rights reserved.
//

#import "PBCommitList.h"
#import "PBGitRevisionCell.h"
#import "PBWebHistoryController.h"
#import "PBHistorySearchController.h"

@implementation PBCommitList

@synthesize mouseDownPoint;
@synthesize useAdjustScroll;

- (NSDragOperation)draggingSourceOperationMaskForLocal:(BOOL) local
{
	return NSDragOperationCopy;
}

- (void)keyDown:(NSEvent *)event
{
	NSString* character = [event charactersIgnoringModifiers];

	// Pass on command-shift up/down to the responder. We want the splitview to capture this.
	if ([event modifierFlags] & NSShiftKeyMask && [event modifierFlags] & NSCommandKeyMask && ([event keyCode] == 0x7E || [event keyCode] == 0x7D)) {
		[self.nextResponder keyDown:event];
		return;
	}

	if ([character isEqualToString:@" "]) {
		if (controller.selectedCommitDetailsIndex == 0) {
			if ([event modifierFlags] & NSShiftKeyMask)
				[webView scrollPageUp:self];
			else
				[webView scrollPageDown:self];
		}
		else
			[controller toggleQLPreviewPanel:self];
	}
	else if ([character rangeOfCharacterFromSet:[NSCharacterSet characterSetWithCharactersInString:@"jkcv"]].location == 0)
		[webController sendKey: character];
	else
		[super keyDown: event];
}

// !!! Andre Berg 20100330: Used from -scrollSelectionToTopOfViewFrom: of PBGitHistoryController
// so that when the history controller udpates the branch filter the origin of the superview gets
// shifted into multiples of the row height. Otherwise the top selected row will always be off by
// a little bit depending on how much the bottom half of the split view is dragged down.
- (NSRect)adjustScroll:(NSRect)proposedVisibleRect {

    //NSLog(@"[%@ %s]: proposedVisibleRect: %@", [self class], _cmd, NSStringFromRect(proposedVisibleRect));
    NSRect newRect = proposedVisibleRect;

    // !!! Andre Berg 20100330: only modify if -scrollSelectionToTopOfViewFrom: has set useAdjustScroll to YES
    // Otherwise we'd also constrain things like middle mouse scrolling.
    if (useAdjustScroll) {
        NSInteger rh = (NSInteger)self.rowHeight;
        NSInteger ny = (NSInteger)proposedVisibleRect.origin.y % (NSInteger)rh;
        NSInteger adj = rh - ny;
        // check the targeted row and see if we need to add or subtract the difference (if there is one)...
        NSRect sr = [self rectOfRow:[self selectedRow]];
        // NSLog(@"[%@ %s]: selectedRow %d, rect: %@", [self class], _cmd, [self selectedRow], NSStringFromRect(sr));
        if (sr.origin.y > proposedVisibleRect.origin.y) {
            // NSLog(@"[%@ %s] selectedRow.origin.y > proposedVisibleRect.origin.y. adding adj (%d)", [self class], _cmd, adj);
            newRect = NSMakeRect(newRect.origin.x, newRect.origin.y + adj, newRect.size.width, newRect.size.height);
        } else if (sr.origin.y < proposedVisibleRect.origin.y) {
            // NSLog(@"[%@ %s] selectedRow.origin.y < proposedVisibleRect.origin.y. subtracting ny (%d)", [self class], _cmd, ny);
            newRect = NSMakeRect(newRect.origin.x, newRect.origin.y - ny , newRect.size.width, newRect.size.height);
        } else {
            // NSLog(@"[%@ %s] selectedRow.origin.y == proposedVisibleRect.origin.y. leaving as is", [self class], _cmd);
        }
    }
    //NSLog(@"[%@ %s]: newRect: %@", [self class], _cmd, NSStringFromRect(newRect));
    return newRect;
}

- (void)mouseDown:(NSEvent *)theEvent
{
    mouseDownPoint = [self convertPoint:[theEvent locationInWindow] fromView:nil];
	[super mouseDown:theEvent];
}

- (NSImage *)dragImageForRowsWithIndexes:(NSIndexSet *)dragRows
							tableColumns:(NSArray *)tableColumns
								   event:(NSEvent *)dragEvent
								  offset:(NSPointPointer)dragImageOffset
{
	NSPoint location = mouseDownPoint;
	NSInteger row = [self rowAtPoint:location];
	NSInteger column = [self columnAtPoint:location];
	PBGitRevisionCell *cell = (PBGitRevisionCell *)[self viewAtColumn:column row:row makeIfNecessary:NO];
	NSRect cellFrame = [self frameOfCellAtColumn:column row:row];

	int index = -1;

	if ([cell respondsToSelector:@selector(indexAtX:)]) {
		index = [cell indexAtX:(location.x - cellFrame.origin.x)];
	}

	if (index == -1)
		return [super dragImageForRowsWithIndexes:dragRows tableColumns:tableColumns event:dragEvent offset:dragImageOffset];

	NSRect rect = [cell rectAtIndex:index];

	NSImage *newImage = [[NSImage alloc] initWithSize:NSMakeSize(rect.size.width + 3, rect.size.height + 3)];
	rect.origin = NSMakePoint(0.5, 0.5);

	[newImage lockFocus];
	[cell drawLabelAtIndex:index inRect:rect];
	[newImage unlockFocus];

	*dragImageOffset = NSMakePoint(rect.size.width / 2 + 10, 0);
	return newImage;

}


#pragma mark Menu

- (NSMenu *)menuForEvent:(NSEvent *)event
{
	[super menuForEvent:event];
	NSInteger index = self.clickedRow;

	NSInteger column = [self columnWithIdentifier:@"SubjectColumn"];
	PBGitRevisionCell *cell = [self viewAtColumn:column row:index makeIfNecessary:NO];
	PBGitCommit *commit = cell.objectValue;

	NSPoint point = [self.window.contentView convertPoint:[event locationInWindow] toView:cell];
	int i = [cell indexAtX:point.x];
	PBGitRef *clickedRef = (i >= 0 ? commit.refs[0] : nil);
	
	NSArray <PBGitCommit*>* selectedCommits = controller.selectedCommits;
	NSArray <NSMenuItem *>* items;

	if (clickedRef) {
		items = [contextMenuDelegate menuItemsForRef:clickedRef];
	} else if ([selectedCommits containsObject:commit]) {
		items = [contextMenuDelegate menuItemsForCommits:controller.selectedCommits];
	} else {
		items = [contextMenuDelegate menuItemsForCommits:@[commit]];
	}

	NSMenu *menu = [[NSMenu alloc] init];
	[menu setAutoenablesItems:NO];
	for (NSMenuItem *item in items)
		[menu addItem:item];

	return menu;
}


#pragma mark Row highlighting

- (IBAction)performFindPanelAction:(id)sender
{
	PBFindPanelActionBlock block = self.findPanelActionBlock;
	if (block) {
		block(sender);
	}
}

@end

File: Classes/PBCommitList.swift
//
//  PBCommitList.swift
//  GitX
//
//  Converted from PBCommitList.m
//

import Cocoa
import WebKit

@objc class PBCommitList: NSTableView {

    // MARK: - Properties

    @IBOutlet weak var webView: WebView!
    @IBOutlet weak var webController: PBWebHistoryController!
    @IBOutlet weak var controller: PBGitHistoryController!
    @IBOutlet weak var searchController: PBHistorySearchController!

    @objc dynamic var useAdjustScroll: Bool = false
    @objc private(set) var mouseDownPoint: NSPoint = .zero

    // MARK: - Lifecycle

    override func awakeFromNib() {
        super.awakeFromNib()
        webView.drawsBackground = false
    }

    // MARK: - Drag and Drop

    override func draggingSession(_ session: NSDraggingSession,
                                   sourceOperationMaskFor context: NSDraggingContext) -> NSDragOperation {
        return .copy
    }

    // TODO: dragImageForRowsWithIndexes is deprecated in modern macOS
    // This custom drag image implementation may need to be reimplemented using modern APIs
    // For now, using the default implementation
    /*
    override func dragImageForRowsWithIndexes(_ dragRows: NSIndexSet,
                                              tableColumns: [NSTableColumn],
                                              event dragEvent: NSEvent,
                                              offset dragImageOffset: NSPointPointer) -> NSImage {
        let location = mouseDownPoint
        let row = self.row(at: location)
        let column = self.column(at: location)

        guard let cell = view(atColumn: column, row: row, makeIfNecessary: false) as? PBGitRevisionCell else {
            return super.dragImageForRowsWithIndexes(dragRows,
                                                     tableColumns: tableColumns,
                                                     event: dragEvent,
                                                     offset: dragImageOffset)
        }

        let cellFrame = frameOfCell(atColumn: column, row: row)
        let index = cell.responds(to: #selector(PBGitRevisionCell.index(atX:)))
            ? cell.index(atX: location.x - cellFrame.origin.x)
            : -1

        if index == -1 {
            return super.dragImageForRowsWithIndexes(dragRows,
                                                     tableColumns: tableColumns,
                                                     event: dragEvent,
                                                     offset: dragImageOffset)
        }

        var rect = cell.rectAtIndex(index)
        let newImage = NSImage(size: NSSize(width: rect.size.width + 3,
                                            height: rect.size.height + 3))
        rect.origin = NSPoint(x: 0.5, y: 0.5)

        newImage.lockFocus()
        cell.drawLabelAtIndex(index, inRect: rect)
        newImage.unlockFocus()

        dragImageOffset.pointee = NSPoint(x: rect.size.width / 2 + 10, y: 0)
        return newImage
    }
    */

    // MARK: - Keyboard Handling

    override func keyDown(with event: NSEvent) {
        guard let character = event.charactersIgnoringModifiers else {
            super.keyDown(with: event)
            return
        }

        let modifiers = event.modifierFlags

        // Pass on command-shift up/down to the responder. We want the splitview to capture this.
        if modifiers.contains(.shift) && modifiers.contains(.command) &&
           (event.keyCode == 0x7E || event.keyCode == 0x7D) {
            nextResponder?.keyDown(with: event)
            return
        }

        if character == " " {
            if controller.selectedCommitDetailsIndex == 0 {
                if modifiers.contains(.shift) {
                    webView.scrollPageUp(self)
                } else {
                    webView.scrollPageDown(self)
                }
            } else {
                controller.toggleQLPreviewPanel(self)
            }
        } else if let range = character.rangeOfCharacter(from: CharacterSet(charactersIn: "jkcv")),
                  range.lowerBound == character.startIndex {
            webController.sendKey(character)
        } else if let firstChar = character.utf16.first,
                  firstChar == NSDownArrowFunctionKey && modifiers.contains(.control) {
            controller.selectParentCommit(self)
        } else {
            super.keyDown(with: event)
        }
    }

    // MARK: - Scroll Adjustment

    override func adjustScroll(_ proposedVisibleRect: NSRect) -> NSRect {
        var newRect = proposedVisibleRect

        // Only modify if scrollSelectionToTopOfViewFrom has set useAdjustScroll to true
        // Otherwise we'd also constrain things like middle mouse scrolling.
        guard useAdjustScroll else {
            return newRect
        }

        let rh = Int(rowHeight)
        let ny = Int(proposedVisibleRect.origin.y) % rh
        let adj = rh - ny

        // Check the targeted row and see if we need to add or subtract the difference
        let sr = rect(ofRow: selectedRow)

        if sr.origin.y > proposedVisibleRect.origin.y {
            newRect = NSRect(x: newRect.origin.x,
                           y: newRect.origin.y + CGFloat(adj),
                           width: newRect.size.width,
                           height: newRect.size.height)
        } else if sr.origin.y < proposedVisibleRect.origin.y {
            newRect = NSRect(x: newRect.origin.x,
                           y: newRect.origin.y - CGFloat(ny),
                           width: newRect.size.width,
                           height: newRect.size.height)
        }

        return newRect
    }

    // MARK: - Mouse Handling

    override func mouseDown(with event: NSEvent) {
        mouseDownPoint = convert(event.locationInWindow, from: nil)
        super.mouseDown(with: event)
    }

    // MARK: - Context Menu

    override func menu(for event: NSEvent) -> NSMenu? {
        super.menu(for: event)
        let index = clickedRow

        let column = self.column(withIdentifier: NSUserInterfaceItemIdentifier("SubjectColumn"))
        guard column != -1,
              let cell = view(atColumn: column, row: index, makeIfNecessary: false) as? PBGitRevisionCell,
              let commit = cell.objectValue as? PBGitCommit else {
            return nil
        }

        let point = window?.contentView?.convert(event.locationInWindow, to: cell) ?? .zero
        let i = Int(cell.indexAt(x: point.x))
        let clickedRef: PBGitRef? = (i >= 0 && i < commit.refs.count) ? commit.refs[i] as? PBGitRef : nil

        let selectedCommits = controller.selectedCommits as? [PBGitCommit] ?? []
        let items: [NSMenuItem]

        if let clickedRef = clickedRef {
            items = controller.menuItems(for: clickedRef) as? [NSMenuItem] ?? []
        } else if selectedCommits.contains(commit) {
            items = controller.menuItems(for: controller.selectedCommits) as? [NSMenuItem] ?? []
        } else {
            items = controller.menuItems(for: [commit]) as? [NSMenuItem] ?? []
        }

        let menu = NSMenu()
        menu.autoenablesItems = false
        for item in items {
            menu.addItem(item)
        }

        return menu
    }
}








